File: src/compilecommands.rs
use std::path::{Path, PathBuf};

use serde::{Serialize, Deserialize};
use structopt::StructOpt;
use std::collections::HashMap;
use thiserror::Error;

use crate::
{
    core,
    printer
};

const COMPILE_COMMANDS_FILE_NAME : &str = "compile_commands.json";

#[derive(Debug)]
pub struct CompileCommand
{
    directory: PathBuf,
    command: String
}


#[derive(Error, Debug)]
enum CompileCommandsError
{
    #[error(transparent)]
    Io(#[from] std::io::Error),

    #[error(transparent)]
    Serialization(#[from] serde_json::error::Error)
}


impl CompileCommand
{
    pub fn get_relative_includes(&self) -> Vec<PathBuf>
    {
        // shitty comamndline parser... beware
        let mut r = Vec::<PathBuf>::new();
        let commands = self.command.split(' ');
        for c in commands
        {
            if let Some(path) = c.strip_prefix("-I")
            {
                let pb = PathBuf::from(path);
                r.push(pb);
            }
        }

        r
    }
}

#[derive(Serialize, Deserialize, Debug)]
struct CompileCommandJson
{
    file: String,
    directory: PathBuf,
    command: String
}


pub fn load_compile_commands(print: &mut printer::Printer, path: &Path) -> HashMap<PathBuf, CompileCommand>
{
    match load_compile_commands_or(path)
    {
        Ok(r) => r,
        Err(e) =>
        {
            print.error(format!("Unable to load compile commands from {}: {}", path.display(), e).as_str());
            HashMap::new()
        }
    }
}

fn load_compile_commands_or(path: &Path) -> Result<HashMap<PathBuf, CompileCommand>, CompileCommandsError>
{
    let content = core::read_file_to_string_x(path)?;
    let data : Result<Vec::<CompileCommandJson>, serde_json::error::Error> = serde_json::from_str(&content);
    let store = data?;

    let mut r = HashMap::new();
    for entry in store
    {
        r.insert
        (
            PathBuf::from(entry.file),
            CompileCommand
            {
                directory: entry.directory,
                command: entry.command
            }
        );
    }
    Ok(r)
}


// find the build folder containing the compile_commands file or None
fn find_build_root(root: &Path) -> Option<PathBuf>
{
    for relative_build in ["build", "build/debug-clang"]
    {
        let build = core::join(root, relative_build);
        let compile_commands_json = core::join(&build, COMPILE_COMMANDS_FILE_NAME);
        if compile_commands_json.exists()
        {
            return Some(build);
        }
    }

    None
}


#[derive(StructOpt, Debug)]
pub struct CompileCommandArg
{
    /// the path to compile_commands.json
    #[structopt(long)]
    compile_commands: Option<PathBuf>
}


impl CompileCommandArg
{
    pub fn get_argument_or_none_with_cwd(&self) -> Option<PathBuf>
    {
        match std::env::current_dir()
        {
            Ok(cwd) => self.get_argument_or_none(&cwd),
            Err(_) => None
        }
    }

    pub fn get_argument_or_none(&self, cwd: &Path) -> Option<PathBuf>
    {
        match &self.compile_commands
        {
            Some(existing) => Some(existing.to_path_buf()),
            None=>
            {
                let mut r = find_build_root(cwd)?;
                r.push(COMPILE_COMMANDS_FILE_NAME);
                Some(r)
            }
        }
    }
}





#[derive(StructOpt, Debug)]
pub struct LinesArg
{
    /// File to list lines in
    #[structopt(parse(from_os_str))]
    filename: PathBuf,

    /// List statements instead
    #[structopt(long)]
    statements: bool,

    /// List blocks instead
    #[structopt(long)]
    blocks: bool
}

/// Tool to list headers
#[derive(StructOpt, Debug)]
pub enum Options
{
    /// list all files in the compile commands struct
    Files
    {
        #[structopt(flatten)]
        cc: CompileCommandArg
    },

    /// list include directories per file
    Includes
    {
        #[structopt(flatten)]
        cc: CompileCommandArg
    }
}

fn handle_files(print: &mut printer::Printer, cc: &CompileCommandArg)
{
    if let Some(path) = cc.get_argument_or_none_with_cwd()
    {
        let commands = load_compile_commands(print, &path);
        print.info(format!("{:#?}", commands).as_str());
    }
}

fn handle_includes(print: &mut printer::Printer, cc: &CompileCommandArg)
{
    if let Some(path) = cc.get_argument_or_none_with_cwd()
    {
        let commands = load_compile_commands(print, &path);
        for (file, command) in &commands
        {

            print.info(format!("{}", file.display()).as_str());
            let dirs = command.get_relative_includes();
            for d in dirs
            {
                print.info(format!("    {}", d.display()).as_str());
            }
        }
    }
}

pub fn main(print: &mut printer::Printer, args: &Options)
{
    match args
    {
        Options::Files{cc} =>
        {
            handle_files(print, cc);
        },
        Options::Includes{cc} =>
        {
            handle_includes(print, cc);
        }
    }
}

File: src/CompileCommands.cs
using Newtonsoft.Json;
using Spectre.Console.Cli;
using System.ComponentModel;
using System.Diagnostics.CodeAnalysis;

namespace Workbench.CompileCommands;

public class CompileCommand
{
    public string directory;
    public string command;

    public CompileCommand(string directory, string command)
    {
        this.directory = directory;
        this.command = command;
    }

    public IEnumerable<string> get_relative_includes()
    {
        // shitty comamndline parser... beware
        foreach (var c in command.Split(' '))
        {
            const string include_prefix = "-I";
            if (c.StartsWith(include_prefix) == false) { continue; }

            yield return c[include_prefix.Length..].Trim();
        }
    }

    public Dictionary<string, string> get_defines()
    {
        // shitty comamndline parser... beware
        var r = new Dictionary<string, string>();

        foreach (var c in command.Split(' '))
        {
            const string define_prefix = "-D";
            if (c.StartsWith(define_prefix) == false) { continue; }

            var def = c[define_prefix.Length..];

            var arr = def.Split('=', 2);
            var key = arr[0];
            var val = arr.Length == 1 ? "" : arr[1];
            r.Add(key, val);
        }

        return r;
    }
}



internal class CompileCommandJson
{
    [JsonProperty("file")]
    public string file = "";

    [JsonProperty("directory")]
    public string directory = "";

    [JsonProperty("command")]
    public string command = "";
}

internal static class Utils
{
    internal static Dictionary<string, CompileCommand>? load_compile_commands(Printer printer, string path)
    {
        var content = File.ReadAllText(path);
        var store = JsonUtil.Parse<List<CompileCommandJson>>(printer, path, content);

        if (store == null)
        {
            printer.error($"Unable to load compile commands from {path}");
            return null;
        }

        var r = new Dictionary<string, CompileCommand>();
        foreach (var entry in store)
        {
            r.Add
            (
                entry.file,
                new CompileCommand
                (
                    directory: entry.directory,
                    command: entry.command
                )
            );
        }

        return r;
    }

    internal const string COMPILE_COMMANDS_FILE_NAME = "compile_commands.json";

    /// find the build folder containing the compile_commands file or None
    public static string? find_build_root(string root)
    {
        var common_roots = new string[] { "build", "build/debug-clang" };

        foreach (var relative_build in common_roots)
        {
            var build = Path.Join(root, relative_build);
            var compile_commands_json = Path.Join(build, COMPILE_COMMANDS_FILE_NAME);
            if (Path.Exists(compile_commands_json))
            {
                return build;
            }
        }

        return null;
    }
}


internal class MainCommandSettings : CommandSettings
{
    [Description("the path to compile_commands.json")]
    [CommandOption("--compile-commands")]
    [DefaultValue(null)]
    string? compile_commands { get; set; }

    public string? get_argument_or_none_with_cwd()
    {
        return get_argument_or_none(Environment.CurrentDirectory);
    }

    public string? get_argument_or_none(string cwd)
    {
        if (compile_commands != null)
        {
            return compile_commands;
        }

        var r = Utils.find_build_root(cwd);
        if (r == null) { return null; }
        return Path.Join(r, Utils.COMPILE_COMMANDS_FILE_NAME);
    }
}


internal sealed class FilesCommand : Command<FilesCommand.Arg>
{
    public sealed class Arg : MainCommandSettings
    {
    }

    public override int Execute([NotNull] CommandContext context, [NotNull] Arg settings)
    {
        return CommonExecute.WithPrinter
            (
                print =>
                {

                    var path = settings.get_argument_or_none_with_cwd();
                    if (path != null)
                    {
                        var commands = Utils.load_compile_commands(print, path);
                        if (commands == null) { return -1; }

                        print.Info($"{commands}");
                    }
                    return 0;
                }
            );
    }
}


internal sealed class IncludesCommand : Command<IncludesCommand.Arg>
{
    public sealed class Arg : MainCommandSettings
    {
    }

    public override int Execute([NotNull] CommandContext context, [NotNull] Arg settings)
    {
        return CommonExecute.WithPrinter
            (
                print =>
                {
                    var path = settings.get_argument_or_none_with_cwd();
                    if (path != null)
                    {
                        var commands = Utils.load_compile_commands(print, path);
                        if (commands == null) { return -1; }

                        foreach (var (file, command) in commands)
                        {
                            print.Info($"{file}");
                            var dirs = command.get_relative_includes();
                            foreach (var d in dirs)
                            {
                                print.Info($"    {d}");
                            }
                        }
                    }
                    return 0;
                }
            );
    }
}

internal sealed class DefinesCommand : Command<DefinesCommand.Arg>
{
    public sealed class Arg : MainCommandSettings
    {
    }

    public override int Execute([NotNull] CommandContext context, [NotNull] Arg settings)
    {
        return CommonExecute.WithPrinter
            (
                print =>
                {
                    var path = settings.get_argument_or_none_with_cwd();
                    if (path != null)
                    {
                        var commands = Utils.load_compile_commands(print, path);
                        if (commands == null) { return -1; }

                        foreach (var (file, command) in commands)
                        {
                            print.Info($"{file}");
                            var defs = command.get_defines();
                            foreach (var (k, v) in defs)
                            {
                                print.Info($"    {k} = {v}");
                            }
                        }
                    }
                    return 0;
                }
            );
    }
}

internal static class Main
{
    internal static void Configure(IConfigurator config, string name)
    {
        config.AddBranch(name, cmake =>
        {
            cmake.SetDescription("Tool to list headers");
            cmake.AddCommand<FilesCommand>("files").WithDescription("list all files in the compile commands class");
            cmake.AddCommand<IncludesCommand>("includes").WithDescription("list include directories per file");
            cmake.AddCommand<DefinesCommand>("defines").WithDescription("list include directories per file");
        });
    }
}




